### ISSUE TITLE
Please consider migrating from pycrypto to pycryptodome

### ISSUE THREAD

**Issue by akavel** [oduwsdl/ipwb#303]
*pycrypto* doesn't currently build on Windows out of the box, requiring [weird workarounds](https://stackoverflow.com/q/41843266/98528), which can be found only after some googling. This is [reported to *pycrypto*](https://github.com/dlitz/pycrypto/issues/221), but then deferred to the following answer:

> Please see https://github.com/dlitz/pycrypto/issues/238.

There, [one can read](https://github.com/dlitz/pycrypto/issues/238):

> **tl;dr:** PyCrypto is dead. Switch to PyCryptodome (GitHub: https://github.com/Legrandin/pycryptodome PyPi: https://pypi.python.org/pypi/pycryptodome). **Do not report any new bugs here!**
> 
> ----
>
> As you may or may not have noticed the last change to PyCrypto was made **three years ago**.
> 
> As a result the developers don't react to issues on the issue tracker anymore. If you find problems and want them to be fixed consider following:
> 
> As a **user** or **developer** who **uses** software which requires PyCrypto:
> 
> -    Tell this to your application's developer to switch to PyCryptodome and tell them PyCrypto is dead.
> 
> As a **developer** who **develops/maintains software depending on PyCrypto**:
> 
>  -   Consider switching to Pycryptodome - it's a well mantained fork and mostly compatible. If your problems persists report your problem to Pycryptodome's developer.
> 
> As a **developer** who wants to **develop new software**:
> 
>   -  _Do not use Pycrypto!_ Use Pycrpytodome or cryptography. PyCrypto is old and not maintained.
> 

Apart from hopefully fixing the issue on Windows, I imagine using a maintained crypto library could be potentially safer?

**Comment by ibnesayeed** [comment]
How's the API compatibility between the two libraries?

**Comment by akavel** [comment]
I don't know anything more about it, just trying to install & run ipwb on Windows and reporting issues I find. Per the 2nd bullet point above:

> "[...] it's a well mantained fork and mostly compatible. If your problems persists report your problem to Pycryptodome's developer."

Gives some hope, especially when accompanied by the request to report problems.

**Comment by ibnesayeed** [comment]
@machawk1 please look into this.

**Comment by machawk1** [comment]
@akavel Thanks for the info. I'll work on moving over to a module other than crypto.

**Comment by machawk1** [comment]
Pycryptodome looks to be mostly a drop-in replacement for pycrypto. However, our rudimentary encryption process uses XOR from Crypto.Cipher, which is not present in the former. The docs say to use Crypto.Util.strxor but the signature is a little different and there's no encrypt() to apply to the cipher.

**Comment by ibnesayeed** [comment]
There is a [CVE-2018-6594](https://nvd.nist.gov/vuln/detail/CVE-2018-6594) affecting PyCrypto.

**Comment by machawk1** [comment]
It has [been reported](https://github.com/dlitz/pycrypto/issues/253) that the alternative we were to use per above, Pycryptrodome, as well as libgcrypt have the same vulnerability.

Luckily our reliance, usage, and (un-luckily) integration with encryption is not thorough or critical. Regardless, this should be a wake-up call to do it right from scratch without too much productivity loss overhead.

**Comment by rililinx** [comment]
CVE-2018-6594 vulnerability is fixed in PyCryptodome 3.4.10 and 3.4.10x by [this](https://github.com/Legrandin/pycryptodome/commit/99c27a3b9e8a884bbde0e88c63234b669d4398d8#diff-f2503cc79ff7f959272896657e0478ac) commit.

**Comment by machawk1** [comment]
@rililinx Thanks for this reference. Other sources also say that it was fixed in the version you mentioned released in Feb 2018. @ibnesayeed Are you ok with the plan to move over to pycryptodome from pycrypto?

**Comment by ibnesayeed** [comment]
I never objected to move to it. Is it a drop-in replacement?

**Comment by ibnesayeed** [comment]
[Compatibility with PyCrypto](http://pycryptodome.readthedocs.io/en/latest/src/vs_pycrypto.html) has XOR functionality documented.

**Comment by machawk1** [comment]
@ibnesayeed I think I may have found the issue -- perhaps the test-requirements.txt file did not include pycryptodome. I am unsure why the tests previously passed without importing pycrypto. I will let them run again to see if this issue is resolved.

**Comment by ibnesayeed** [comment]
While including the package in the test requirements is important, the functionality will perhaps not work until we replace `Crypto.Cipher.XOR` with `Crypto.Util.strxor`.

**Comment by machawk1** [comment]
I was having issue replicating what TravisCI said re:tests for this ticket. I still had pycrypto installed and thus the `XOR` function was still available.

**Comment by ibnesayeed** [comment]
> I still had pycrypto installed and thus the XOR function was still available.

Did I mention a thing called Docker? ;-)

**Comment by machawk1** [comment]
It appears that the `Crypto.Util.strxor` part of the API does not have what we need, namely it appears that there is not an correlative `decrypt()` function. This is problematic, as we need to be able to decrypt the values using the same secret key. I will need to run some tests and read the docs a bit more to determine if there is another means to perform this same function in pycryptodome.

**Comment by ibnesayeed** [comment]
I have not looked into that part of the code, but shouldn't it be the case that we can apply the very same XOR method with the same key to decrypt the message?

**Comment by machawk1** [comment]
It looks like the [conventional ciphers of pycryptodome](http://pycryptodome.readthedocs.io/en/latest/src/cipher/classic.html) still support the same `new()` instantiation and encrypt/decrypt APIs.

**Comment by machawk1** [comment]
Caveat (discussed in-person): key and data to be encrypted must be same length for XOR. Padding the key using `Crypto.Util.Padding.pad()` _should_ work but I get errors like the below: 
```
  File "/private/tmp/ipwb/xx/lib/python2.7/site-packages/Crypto/Util/Padding.py", line 57, in pad
    padding = bchr(padding_len)*padding_len
  File "/private/tmp/ipwb/xx/lib/python2.7/site-packages/Crypto/Util/py3compat.py", line 67, in bchr
    return chr(s)
ValueError: chr() arg not in range(256)
```
...when running `strxor(pad(key, len(data)), data)`.

**Comment by ibnesayeed** [comment]
XOR is too primitive, use something like AES instead. Alternatively, you can segment the data in smaller blocks and encrypt them separately then combine, but then the reverse operation needs to know how you segmented initially.
